hist/histv7/inc/ROOT/RHistAutoAxisFiller.hxx

```diff
@@ -0,0 +1,195 @@
+   std::size_t fMaxBufferSize;
+
+   using BufferElement = std::conditional_t<SupportsWeightedFilling, std::pair<double, RWeight>, double>;
+
+   /// The buffer of filled entries
+   std::vector<BufferElement> fBuffer;
+   /// The minimum of the filled entries
+   double fMinimum = std::numeric_limits<double>::infinity();
+   /// The maximum of the filled entries
+   double fMaximum = -std::numeric_limits<double>::infinity();
+
+public:
+   /// Create a filler object.
+   ///
+   /// \param[in] nNormalBins the number of normal bins, must be > 0
+   /// \param[in] maxBufferSize the maximum buffer size, must be > 0
+   explicit RHistAutoAxisFiller(std::size_t nNormalBins, std::size_t maxBufferSize = 1024)
+      : fNNormalBins(nNormalBins), fMaxBufferSize(maxBufferSize)
+   {
+      if (nNormalBins == 0) {
+         throw std::invalid_argument("nNormalBins must be > 0");
+      }
+      if (maxBufferSize == 0) {
+         throw std::invalid_argument("maxBufferSize must be > 0");
+      }
+   }
+
+   std::size_t GetNNormalBins() const { return fNNormalBins; }
+   std::size_t GetMaxBufferSize() const { return fMaxBufferSize; }
+
+private:
+   void BufferImpl(double x, RWeight weight)
+   {
+      if constexpr (SupportsWeightedFilling) {
+         fBuffer.emplace_back(x, weight);
+      } else {
+         assert(weight.fValue == 1.0);
+         // Silence compiler warning about unused parameter
+         (void)weight;
+         fBuffer.push_back(x);
+      }
+      if (x < fMinimum) {
+         fMinimum = x;
+      }
+      if (x > fMaximum) {
+         fMaximum = x;
+      }
+
+      if (fBuffer.size() >= fMaxBufferSize) {
+         Flush();
+      }
+   }
+
+public:
+   /// Fill an entry into the histogram.
+   ///
+   /// \param[in] x the argument
+   /// \par See also
+   /// the \ref Fill(double x, RWeight weight) "overload for weighted filling"
+   void Fill(double x)
+   {
+      // If the histogram exists, forward the Fill call.
+      if (fHist) {
+         fHist->Fill(x);
+         return;
+      }
+      BufferImpl(x, RWeight(1.0));
+   }
+
+   /// Fill an entry into the histogram with a weight.
+   ///
+   /// This overload is only available for floating-point bin content types (see
+   /// \ref RHistEngine::SupportsWeightedFilling).
+   ///
+   /// \param[in] x the argument
+   /// \param[in] weight the weight for this entry
+   /// \par See also
+   /// the \ref Fill(double x) "overload for unweighted filling"
+   void Fill(double x, RWeight weight)
+   {
+      // If the histogram exists, forward the Fill call.
+      if (fHist) {
+         fHist->Fill(x, weight);
+         return;
+      }
+      BufferImpl(x, weight);
+   }
+
+   /// Flush the buffer of entries and construct the histogram.
+   ///
+   /// Throws an exception if the buffer is empty, the axis interval cannot be determined, or if it would be empty
+   /// because the minimum equals the maximum.
+   void Flush()
```

Comment:
Do we want/need to expose this publicly?

Reply:
Another good question. A potential use case would be where the user designates a limited set of "sample" entries that are potentially not aligned with `fMaxBufferSize`, or multiple `RHistAutoAxisFiller` with different fill frequencies that the user wants to "synchronize".